tests/trainer/test_tp.py

```diff
@@ -1,27 +1,456 @@
+    compare_models(ddp_trainer, fsdp_trainer, tp_fsdp_trainer)
+
+    # Forward pass
+    ddp_out = forward_pass(ddp_trainer)
+    fsdp_out = forward_pass(fsdp_trainer)
+    tp_fsdp_out = forward_pass(tp_fsdp_trainer)
+
+    # Ensure output of the forward pass is the same
+    with FSDP.summon_full_params(fsdp_trainer.state.model):
+        with FSDP.summon_full_params(tp_fsdp_trainer.state.model):
+            compare_modules({'': ddp_out}, {'': fsdp_out})
+            compare_modules({'': ddp_out}, {'': tp_fsdp_out})
+            compare_modules({'': fsdp_out}, {'': tp_fsdp_out})
+
+    # Compute gradients
+    torch.sum(ddp_out).backward()
+    torch.sum(fsdp_out).backward()
+    torch.sum(tp_fsdp_out).backward()
+
+    # Ensure the gradients are the same
+    compare_models(ddp_trainer, fsdp_trainer, tp_fsdp_trainer, check_grad=True)
+
+    # Update the model weights
+    ddp_trainer.state.optimizers[0].step()
+    fsdp_trainer.state.optimizers[0].step()
+    tp_fsdp_trainer.state.optimizers[0].step()
+
+    # Ensure the updated weights are the same
+    compare_models(ddp_trainer, fsdp_trainer, tp_fsdp_trainer)
+
+
+@pytest.mark.gpu
+@world_size(4)
+@pytest.mark.parametrize('replication', [2])
+@pytest.mark.parametrize('batch_size', [1, 4])
+@pytest.mark.skipif(version.parse(torch.__version__) < version.parse('2.3'), reason='Requires PyTorch 2.3+')
+@pytest.mark.filterwarnings(r'ignore:.*\(TP\) is experimental.*:FutureWarning')
+def test_tp_fit_correctness(world_size: int, batch_size: int, replication: int):
+    """Test that training with DDP, FSDP, TP-FSDP results in the same:
+        - updated weights
+        - loss
+        - accuracy
+    after training for multiple steps via trainer.fit().
+    """
+
+    # Initialize number of samples in the dataset
+    # train_steps = 20  # number of steps to train for
+    train_steps = 20
+    samples_per_batch = world_size * batch_size // replication
+    dataset_size = samples_per_batch * train_steps
+
+    # DDP fit
+    ddp_trainer = get_ddp_trainer(size=dataset_size, batch_size=batch_size, replication=replication)
+    ddp_trainer.fit()
+    ddp_trainer.close()
+    ddp_stats = get_stats(ddp_trainer)
+
+    # FSDP fit
+    fsdp_trainer = get_fsdp_trainer(size=dataset_size, batch_size=batch_size, replication=replication)
+    fsdp_trainer.fit()
+    fsdp_trainer.close()
+    fsdp_stats = get_stats(fsdp_trainer)
+
+    # TP-FSDP fit
+    tp_fsdp_trainer = get_tp_fsdp_trainer(size=dataset_size, batch_size=batch_size, replication=replication)
+    tp_fsdp_trainer.fit()
+    tp_fsdp_trainer.close()
+    tp_fsdp_stats = get_stats(tp_fsdp_trainer)
+
+    # Ensure the updated models weights are the same
+    # Drop tolerance due to precision issues across different parallelism strategies
+    compare_models(ddp_trainer, fsdp_trainer, tp_fsdp_trainer, atol=1e-5, rtol=1e-3)
+
+    # Compare loss between DDP, FSDP, TP-FSDP
+    np.testing.assert_allclose(
```

Comment:
why np and not torch?

Reply:
because the accuracy is saved as a np array by default in the in-memory logger